File: src/WetherSubject.dart
import 'ObserverInterface.dart';
import 'SubjectInterface.dart';

class WetherSubject implements SubjectInterface {
  late List<ObserverInterface> _observers;
  late double _temperature;

  WetherSubject() {
    _observers = [];
  }

  void registerObserver(ObserverInterface observer) {
    _observers.add(observer);
  }

  void removeObserver(ObserverInterface observer) {
    _observers.removeWhere((element) => element == observer);
  }

  void notifyObservers() {
    _observers.forEach((observer) {
      if (_temperature != Null) observer.update(_temperature);
    });
  }

  void setTemperature(double temperature) {
    _temperature = temperature;
    this.notifyObservers();
  }
}

File: src/WetherSubject.ts
import { ObserverInterface } from "./ObserverInterface";
import { SubjectInterface } from "./SubjectInterface";

export class WetherSubject implements SubjectInterface {
  private observers: ObserverInterface[];

  private temperature?: number;

  constructor() {
    this.observers = [];
  }

  registerObserver(observer: ObserverInterface): void {
    this.observers.push(observer);
  }

  removeObserver(observer: ObserverInterface): void {
    const index = this.observers.indexOf(observer);
    if (index >= 0) this.observers.splice(index, 1);
  }

  notifyObservers(): void {
    this.observers.forEach((observer) => {
      if (this.temperature != null) observer.update(this.temperature);
    });
  }

  setTemperature(temperature: number): void {
    this.temperature = temperature;
    this.notifyObservers();
  }
}
